File: house-robber-ii.java
class Solution {
    public int rob(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        if (nums.length == 1) {
            return nums[0];
        }
        return Math.max(rob(nums, 0, nums.length - 1), rob(nums, 1, nums.length));
    }
    
    public int rob(int[] nums, int start, int end) {
        int prevMax = 0;
        int currMax = 0;
        for (int i = start; i < end; i++) {
            int num = nums[i];
            // prevMax + num => take curr
            // currMax => do not take curr
            int temp = currMax;
            currMax = Math.max(prevMax + num, currMax);
            prevMax = temp;
        }
        
        return currMax;
    }
}

File: house-robber-ii/house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        
        return max(self.rob_simple(nums, 0, len(nums) - 2), self.rob_simple(nums, 1, len(nums) - 1))
    
    def rob_simple(self, nums, start, end):
        t1 = 0
        t2 = 0
        for i in range(start, end + 1):
            temp = t1
            t1 = max(t2 + nums[i], t1)
            t2 = temp
        
        return t1